db-migrations/000007_adding_functions.up.sql

```diff
@@ -0,0 +1,275 @@
+                t_valid       BOOL,
+                p_login       VARCHAR,
+                p_email       VARCHAR,
+                p_name        VARCHAR,
+                p_valid       BOOL
+            )
+AS
+$$
+DECLARE
+    var_account_id INTEGER;
+BEGIN
+    CREATE TEMP TABLE IF NOT EXISTS temp_hostname_table AS
+    SELECT ia.public_ip,
+           ia.aws_hostname,
+           res.arn_id,
+           res.meta,
+           ar.region,
+           rt.resource_type,
+           aa.account,
+           aa.id
+    FROM aws_public_ip_assignment ia
+             LEFT JOIN aws_resource res ON ia.aws_resource_id = res.id
+             LEFT JOIN aws_region ar ON res.aws_region_id = ar.id
+             LEFT JOIN aws_resource_type rt ON res.aws_resource_type_id = rt.id
+             LEFT JOIN aws_account aa ON res.aws_account_id = aa.id
+    WHERE ia.aws_hostname = name
+      AND ia.not_before < ts
+      AND (ia.not_after IS NULL OR ia.not_after > ts);
+
+    SELECT temp_hostname_table.id INTO var_account_id FROM temp_hostname_table FETCH FIRST ROW ONLY;
+
+    RETURN QUERY SELECT a.*, b.*
+                 FROM temp_hostname_table AS a
+                          LEFT JOIN (
+                     SELECT *
+                     FROM get_owner_and_champions_by_account_id(var_account_id)
+                 ) AS b ON a.account = b.t_account;
+END;
+$$
+    LANGUAGE 'plpgsql';
+
+CREATE OR REPLACE FUNCTION get_resource_by_private_ip(pip INET, ts TIMESTAMP)
+    RETURNS TABLE
+            (
+                private_ip    INET,
+                arn_id        VARCHAR,
+                meta          JSONB,
+                region        VARCHAR,
+                resource_type VARCHAR,
+                account       VARCHAR,
+                id            INTEGER,
+                t_account     VARCHAR,
+                t_login       VARCHAR,
+                t_email       VARCHAR,
+                t_name        VARCHAR,
+                t_valid       BOOL,
+                p_login       VARCHAR,
+                p_email       VARCHAR,
+                p_name        VARCHAR,
+                p_valid       BOOL
+            )
+AS
+$$
+DECLARE
+    var_account_id INTEGER;
+BEGIN
+    CREATE TEMP TABLE IF NOT EXISTS temp_private_ip_table AS
+    SELECT ia.private_ip,
+           res.arn_id,
+           res.meta,
+           ar.region,
+           rt.resource_type,
+           aa.account,
+           aa.id
+    FROM aws_private_ip_assignment ia
+             LEFT JOIN aws_resource res ON ia.aws_resource_id = res.id
+             LEFT JOIN aws_region ar ON res.aws_region_id = ar.id
+             LEFT JOIN aws_resource_type rt ON res.aws_resource_type_id = rt.id
+             LEFT JOIN aws_account aa ON res.aws_account_id = aa.id
+    WHERE ia.private_ip = pip
+      AND ia.not_before < ts
+      AND (ia.not_after IS NULL OR ia.not_after > ts);
+
+    SELECT temp_private_ip_table.id INTO var_account_id FROM temp_private_ip_table FETCH FIRST ROW ONLY;
```

Comment:
I'm new to sql functions so perhaps I'm misunderstanding something but does selecting only the first accountId mean we can safely make the assumption that if the temp_private_ip_table has multiple results that the accountId in each record will be the same as the first? 
I know it is rare for us to see multiple results in these queries, perhaps it is true that the accountId will be the same for all ip addresses, but I wanted to make sure.

Reply:
This is a great question, i had similar concern while implementation. The existing implementation relied on time stamp `ts` and assumption/reality is there won't be any over lapping `start` and `end` ts. This gives us unique `aws_resource_id` which in turn point to unique `aws_account`.

However, we can certainly check for violation of uniqueness assumption and throw exception.

Reply:
Updated query to incorporate above regression. Also, added test data related to above scenario in latest commit.